File: xy10016/inventory/models/audit_log.py
from sqlalchemy import Column, Integer, String, DateTime, Text
from datetime import datetime
from .base import Base


class AuditLog(Base):
    __tablename__ = 'audit_log'

    action = Column(String(50), nullable=False)
    resource_type = Column(String(50), nullable=False)
    resource_id = Column(String(100))
    user_id = Column(String(50))
    username = Column(String(50))
    ip_address = Column(String(50))
    timestamp = Column(DateTime, default=datetime.utcnow, nullable=False)
    old_value = Column(Text)
    new_value = Column(Text)
    status = Column(String(20), default='success')
    error_message = Column(Text)
    correlation_id = Column(String(100))

    def __repr__(self):
        return f'<AuditLog {self.action} {self.resource_type} at {self.timestamp}>'

File: xy10179/src/models/AuditLog.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');
const Tenant = require('./Tenant');
const User = require('./User');
const DataRecord = require('./DataRecord');

const AuditLog = sequelize.define('AuditLog', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true
  },
  tenantId: {
    type: DataTypes.UUID,
    allowNull: false
  },
  userId: {
    type: DataTypes.UUID,
    allowNull: true
  },
  recordId: {
    type: DataTypes.UUID,
    allowNull: true
  },
  action: {
    type: DataTypes.ENUM('CREATE', 'READ', 'UPDATE', 'DELETE', 'LOGIN', 'LOGOUT', 'TENANT_SWITCH', 'UNAUTHORIZED_ACCESS'),
    allowNull: false
  },
  resourceType: {
    type: DataTypes.STRING(50),
    allowNull: false
  },
  resourceId: {
    type: DataTypes.UUID,
    allowNull: true
  },
  beforeState: {
    type: DataTypes.TEXT,
    allowNull: true
  },
  afterState: {
    type: DataTypes.TEXT,
    allowNull: true
  },
  ipAddress: {
    type: DataTypes.STRING(45),
    allowNull: true
  },
  userAgent: {
    type: DataTypes.STRING(500),
    allowNull: true
  },
  status: {
    type: DataTypes.ENUM('success', 'failure', 'warning'),
    defaultValue: 'success'
  },
  errorMessage: {
    type: DataTypes.TEXT,
    allowNull: true
  },
  requestId: {
    type: DataTypes.STRING(100),
    allowNull: true
  }
}, {
  updatedAt: false,
  indexes: [
    {
      fields: ['tenantId', 'createdAt']
    },
    {
      fields: ['userId', 'action', 'createdAt']
    },
    {
      fields: ['recordId', 'createdAt']
    }
  ]
});

module.exports = AuditLog;
